### ISSUE TITLE
XML specified image types

### ISSUE THREAD

**Issue by h1262216** [h1262216/crazys-wm-mod#16]
I would like to move away from having a hardcoded enum and fallback function for the different image types specified by an xml file. Then we can also easily provide more specialized image types for different actions, and let them fall back to the current images if the target is not present in the pack.

Here are a few points to consider:
1) Ideally, I would like each pack to have an xml which specifies which files to use for which image category, instead of having this implicit in the file name. Given the huge amount of existing packs, we cannot expect file name based tagging to go away anytime soon. Thus, we need at least a basic fallback mechanism in C++, and we should supply scripts that generate xml files based on the file names in a given directory (and maybe vice versa). 

2) Currently, the image type carries multiple pieces of information, which are at least the "Action" being performed (the basic image type, e.g. formal or maid) and some additional tagging (e.g. all the PREG... variants of all basic images). Some of the tagging is currently done in the fallback mechanism, e.g. the profile image might be replaced with something else depending on the girls traits or items.

**Comment by UseAfterFree-177** [comment]
One idea that I've been having would be to have this rich tag format as XML or JSON or YAML (either one metadata file per imagepack or one per picture, ideally we'd let the pack builder choose); and then as a fallback, synthesize our rich tags out of the tags-in-name we've got now.

**Comment by h1262216** [comment]
yes, one metadata file per pack was my original idea, though I guess one per picture would also work. This would make a difference if you want to specify two girls which use the same picture for different things.

**Comment by UseAfterFree-177** [comment]
Trivial case: Metadata lives in the filename, as it does today.

Simple case: The metadata file is named after the picture file, and implicitly refers to it (`Aerith-1.jpg` and `Aerith-1.jpg.xml`). There's a one-to-one relation between picture and metadata. I'd expect this case to cover most pictures.

Advanced case: A metadata field refers to the associated picture file. This would allow for at least many-to-one relations (Ex. `Midna-Barkeep.xml` and `Zelda-Singer.xml` both referring `Zelda-and-Midna.jpg`). 

Optionally: A metadata file that describes the image pack itself (version, short description, download URL, curator's forum handle, etc) .

**Comment by h1262216** [comment]
At this point I am actually more concerned about what metadata to associate with each image, and how to implement the image selection process. How we get this data into C++ is secondary, though I think it is a good idea to have more than one possibility, to give pack makers a bit of freedom. 

But what data do we actually need?
My current model would be something like this:
Each image is associated with a list of use cases. A use case consists of a (primary) action, a set of [secondary] conditions, and an image to use. When we need an image for a given action, we first scan all use cases of the action and check their conditions. If at least one remains, we choose randomly from the valid use cases. The interesting thing is what happens if we don't find an image. Then we have to relax our search criterion. We could:
1) Count how many conditions are violated, and choose among those with the least violated conditions. If no image exists for the action do we start looking through the fallback actions. 
2) Directly start looking through the fallback actions until we arrive at some point where we have an allowed use case. Note that part of the fallback process could also be to search for the same action type, but ignoring certain conditions.
3) Do something smarter?

**Comment by UseAfterFree-177** [comment]
The easy parts would be the exact matches and our default stock pictures. So let's start with those. (This is roughly what we've got today.)

Then we can add in the near-matches in between. With a scoring function we could sort the pictures from good to bad matches, and with some kind of cutoff rule we can then take the best few near-matches. I'd expect the scoring function and the cutoff rules to need a lot of trial-and-error and tuning until we get it roughly right. (As an aside, I believe that this scheme would be able to emulate your cases 1 and 2 -- of course, it would be seriously overdesigned if we actually stopped there.)

Maybe we should sketch out these ideas in some standalone script language like Perl or Python (or whatever). This would let us invent whatever metadata we need to make the scheme work, and when we're done we'd see what we need to bring into C++.
